pkg/metrics_exporter.go

```diff
@@ -0,0 +1,234 @@
+		Handler:           mux,
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
+	return recorder
+}
+
+type noopMetricsRecorder struct{}
+
+func (noopMetricsRecorder) Start(context.Context) {}
+
+func (noopMetricsRecorder) RecordPodEviction(_, _, _ string) {}
+
+func (noopMetricsRecorder) RecordObservedNamespace(_, _, _ string) {}
+
+type prometheusMetricsRecorder struct {
+	server  *http.Server
+	counter *evictedPodCounter
+	once    sync.Once
+}
+
+func (r *prometheusMetricsRecorder) Start(ctx context.Context) {
+	if r.server == nil {
+		return
+	}
+
+	r.once.Do(func() {
+		go func() {
+			<-ctx.Done()
+
+			shutdownCtx, cancel := context.WithTimeout(context.Background(), metricsShutdownGrace)
+			defer cancel()
+
+			if err := r.server.Shutdown(shutdownCtx); err != nil && !errors.Is(err, http.ErrServerClosed) {
+				klog.Errorf("unable to shutdown metrics exporter: %v", err)
+			}
+		}()
+
+		go func() {
+			klog.Infof("Prometheus exporter listening on %s%s", r.server.Addr, metricsPath)
+			if err := r.server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+				klog.Errorf("Prometheus exporter failed: %v", err)
+			}
+		}()
+	})
+}
+
+func (r *prometheusMetricsRecorder) RecordPodEviction(namespace, appName, appInstance string) {
+	if r.counter == nil {
+		return
+	}
+
+	r.counter.Inc(namespace, appName, appInstance)
+}
+
+func (r *prometheusMetricsRecorder) RecordObservedNamespace(namespace, appName, appInstance string) {
+	if r.counter == nil {
+		return
+	}
+
+	r.counter.EnsureMetric(namespace, appName, appInstance)
+}
+
+func (r *prometheusMetricsRecorder) handleMetrics(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodGet && req.Method != http.MethodHead {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+	w.Header().Set("Content-Type", "text/plain; version=0.0.4")
+	w.Header().Set("Cache-Control", "no-store")
+
+	// HEAD only needs headers
+	if req.Method == http.MethodHead {
+		return
+	}
+
+	if err := writeMetrics(w, r.counter); err != nil {
+		klog.Errorf("unable to write metrics response: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func writeMetrics(w io.Writer, counter *evictedPodCounter) error {
+	if counter == nil {
+		return nil
+	}
+
+	if _, err := fmt.Fprintf(w, "# HELP %s %s\n", evictionMetricName, evictionMetricHelp); err != nil {
+		return err
+	}
+
+	if _, err := fmt.Fprintf(w, "# TYPE %s counter\n", evictionMetricName); err != nil {
+		return err
+	}
+
+	return counter.WriteTo(w)
+}
+
+type evictedPodCounter struct {
+	mu     sync.RWMutex
+	counts map[metricKey]uint64
+}
+
+type metricKey struct {
+	namespace string
+	appName   string
+	appInst   string
+}
+
+type metricEntry struct {
+	key   metricKey
+	value uint64
+}
+
+func newEvictedPodCounter() *evictedPodCounter {
+	return &evictedPodCounter{
+		counts: make(map[metricKey]uint64),
+	}
+}
+
+func (c *evictedPodCounter) Inc(namespace, appName, appInstance string) {
```

Comment:
Unless there is a specific need for managing this by hand, have a look at this guide: https://prometheus.io/docs/guides/go-application/

For classic counters, there is a much easier way to achieve this. You could remove the noopRecorder even, and just start the metric server conditionally.